solver/jobs.go

```diff
@@ -766,6 +768,13 @@ func (s *sharedOp) Exec(ctx context.Context, inputs []Result) (outputs []Result,
 		if s.execRes != nil || s.execErr != nil {
 			return s.execRes, s.execErr
 		}
+		if s.st.opts.ParallelismSem != nil {
```

Comment:
There is a case with `BuildOp` where I think this can potentially deadlock. I think we need to detect it somehow and let `BuildOp.Exec` ignore the semaphore.

Reply:
This should be done by extending the `Op` interface, eg in `ResolveOp` or passing something to `op.Exec`. We can't do llb specific type checks in the main solver pkg.